channeldb/migration27/migration.go

```diff
@@ -0,0 +1,123 @@
+package migration27
+
+import (
+	"bytes"
+	"fmt"
+
+	mig "github.com/lightningnetwork/lnd/channeldb/migration_01_to_11"
+	"github.com/lightningnetwork/lnd/kvdb"
+)
+
+var (
+	// historicalChannelBucket stores all channels that have seen their
+	// commitment tx confirm. All information from their previous open state
+	// is retained.
+	historicalChannelBucket = []byte("historical-chan-bucket")
+)
+
+// MigrateHistoricalBalances patches the two new fields, `InitialLocalBalance`
+// and `InitialRemoteBalance`, for all the open channels saved in historical
+// channel bucket. Unlike migration 25, it will only read the old channel info
+// first and then patch the new tlv records with empty values. For historical
+// channels, we previously didn't save the initial balances anywhere and since
+// it's corresponding open channel bucket is deleted after closure, we have
+// lost that balance info.
+func MigrateHistoricalBalances(tx kvdb.RwTx) error {
+	log.Infof("Migrating historical local and remote balances...")
+
+	// First fetch the top level bucket which stores all data related to
+	// historically stored channels.
+	rootBucket := tx.ReadWriteBucket(historicalChannelBucket)
+
+	// If no bucket is found, we can exit early.
+	if rootBucket == nil {
+		return nil
+	}
+
+	// Read a list of historical channels.
+	channels, err := findHistoricalChannels(rootBucket)
+	if err != nil {
+		return err
+	}
+
+	// Migrate the balances.
+	for _, c := range channels {
+		if err := migrateBalances(rootBucket, c); err != nil {
+			return err
+		}
+	}
+
+	return err
+}
+
+// findHistoricalChannels finds all historical channels.
+func findHistoricalChannels(historicalBucket kvdb.RBucket) ([]*OpenChannel,
+	error) {
+
+	channels := []*OpenChannel{}
+
+	// readChannel is a helper closure that reads the channel info from the
+	// historical sub-bucket.
+	readChannel := func(rootBucket kvdb.RBucket, cp []byte) error {
+		c := &OpenChannel{}
+
+		chanPointBuf := bytes.NewBuffer(cp)
+		err := mig.ReadOutpoint(chanPointBuf, &c.FundingOutpoint)
+		if err != nil {
+			return fmt.Errorf("read funding outpoint got: %v", err)
+		}
+
+		// Read the sub-bucket.
+		chanBucket := rootBucket.NestedReadBucket(cp)
+		if chanBucket != nil {
```

Comment:
`!=`?

Reply:
Ah, `chanBucket` contains the error, usually called `err`?

Reply:
nope that was a mistake, I was testing the error logs and forgot to put it back. `chanBucket` is the bucket returned.